analytic_engine/src/sst/parquet/row_group_filter.rs

```diff
@@ -0,0 +1,131 @@
+// Copyright 2022 CeresDB Project Authors. Licensed under Apache-2.0.
+
+// Filter for row groups.
+
+use std::cmp::Ordering;
+
+use arrow::datatypes::SchemaRef;
+use common_types::datum::Datum;
+use datafusion::{prelude::Expr, scalar::ScalarValue};
+use ethbloom::{Bloom, Input};
+use parquet::file::metadata::RowGroupMetaData;
+use parquet_ext::prune::{
+    equal::{self, ColumnPosition},
+    min_max,
+};
+use snafu::ensure;
+
+use crate::sst::reader::error::{OtherNoCause, Result};
+
+/// A filter to prune row groups according to the provided predicates.
+///
+/// Currently, two kinds of filters will be applied to such filtering:
+/// min max & bloom filter.
+pub struct RowGroupFilter<'a> {
+    schema: &'a SchemaRef,
+    row_groups: &'a [RowGroupMetaData],
+    blooms: &'a [Vec<Bloom>],
+    predicates: &'a [Expr],
+}
+
+impl<'a> RowGroupFilter<'a> {
+    pub fn try_new(
+        schema: &'a SchemaRef,
+        row_groups: &'a [RowGroupMetaData],
+        blooms: &'a [Vec<Bloom>],
+        predicates: &'a [Expr],
+    ) -> Result<Self> {
+        ensure!(blooms.len() == row_groups.len(), OtherNoCause {
+            msg: format!("expect the same number of bloom filter as the number of row groups, num_bloom_filters:{}, num_row_groups:{}", blooms.len(), row_groups.len()),
+        });
+
+        Ok(Self {
+            schema,
+            row_groups,
+            blooms,
+            predicates,
+        })
+    }
+
+    pub fn filter(&self) -> Vec<usize> {
+        let filtered0 = self.filter_by_min_max();
+        let filtered1 = self.filter_by_bloom();
```

Comment:
Why not pass filtered0 to `filter_by_bloom`, it will reduce total row groups to filter

Reply:
It's a good idea, but current the interface of `EqPruner` seems not flexible to allow us to such improvement. I will leave a TODO here.